src/loaders/jobs/taskfarmer/task_generator.py

```diff
@@ -0,0 +1,284 @@
+CHECKM2_DB = '/global/cfs/cdirs/kbase/collections/libraries/CheckM2_database'
+
+
+def _run_command(command, job_dir, log_file_prefix='', check_return_code=True):
+    """
+    Runs the specified command and captures its standard output and standard error.
+    """
+    log_dir = os.path.join(job_dir, 'logs')
+    os.makedirs(log_dir, exist_ok=True)
+    std_out_file = os.path.join(log_dir,
+                                f'stdout_{log_file_prefix}_{datetime.datetime.now().strftime("%Y%m%d_%H%M%S")}')
+    std_err_file = os.path.join(log_dir,
+                                f'stderr_{log_file_prefix}_{datetime.datetime.now().strftime("%Y%m%d_%H%M%S")}')
+
+    with open(std_out_file, "w") as std_out, open(std_err_file, "w") as std_err:
+        p = subprocess.Popen(command, stdout=std_out, stderr=std_err, text=True)
+
+    exit_code = p.wait()
+
+    if check_return_code and exit_code != 0:
+        with open(std_out_file, "r") as std_out, open(std_err_file, "r") as std_err:
+            raise ValueError(f"Error running command '{command}'.\n"
+                             f"Standard output: {std_out.read()}\n"
+                             f"Standard error: {std_err.read()}")
+
+    return std_out_file, std_err_file, exit_code
+
+
+def _clone_repo(job_dir):
+    """
+    Clones the collections repo into the specified directory.
+    """
+    repo_dir = os.path.join(job_dir, 'collections')
+    os.rmdir(repo_dir) if os.path.exists(repo_dir) else None  # remove existing repo otherwise git clone will fail
+    _run_command(["git", "clone", REPO_URL, repo_dir], job_dir, log_file_prefix='git_clone')
+
+    if not os.path.exists(repo_dir):
+        raise ValueError(f"Error cloning repo '{REPO_URL}' into '{repo_dir}'.")
+
+
+def _fetch_image(registry, image_name, job_dir, ver='latest'):
+    """
+    Fetches the specified Shifter image, if it's not already present on the system.
+    """
+    image_str = f"{registry}/{image_name}:{ver}"
+
+    # Check if the image is already present on the system
+    std_out_file, std_err_file, exit_code = _run_command(["shifterimg", "images"], job_dir,
+                                                         log_file_prefix='shifterimg_images')
+
+    with open(std_out_file, "r") as f:
+        std_out = f.read()
+
+    images = std_out.split("\n")
+    for image in images:
+        parts = image.split()
+        if len(parts) != 6:
+            continue
+        if parts[5] == image_str:
+            print(f"Shifter image {image_name}:{ver} from registry {registry} already exists.")
+            return parts[5]
+
+    # Pull the image from the registry
+    print(f"Fetching Shifter image {image_str}...")
+    std_out_file, std_err_file, exit_code = _run_command(["shifterimg", "pull", image_str], job_dir,
+                                                         log_file_prefix='shifterimg_pull')
+
+    if 'FAILURE' in std_out:
```

Comment:
This is referring to the std_out from the `images` command

Reply:
👍  Fixed.

Reply:
:+1: